src/plugins/lua/replies.lua

```diff
@@ -91,16 +97,146 @@ local function replies_check(task)
+              { mgs = settings['max_global_size'] })
+    table.insert(script_ids, lua_redis.add_redis_script(set_script_zadd_global, redis_params))
+
+    local redis_script_zadd_local = [[
+      local local_replies_set_size = redis.call('ZCARD', KEYS[1])
+      if local_replies_set_size > {= mls =} then
+        redis.call('ZREMRANGEBYRANK', KEYS[1], '{= mls =}', local_replies_set_size)
+      end
+      -- adding recipients for the local replies set
+      redis.call('ZADD', KEYS[1], ARGV[1], ARGV[2])
+      -- setting expire for local replies set
+      redis.call('EXPIRE', KEYS[1], tostring(math.floor('{= exp =}')))
+    ]]
+    local set_script_zadd_local = lua_util.jinja_template(redis_script_zadd_local,
+      { exp = settings['expire'], mls = settings['max_local_size'] })
+    table.insert(script_ids, lua_redis.add_redis_script(set_script_zadd_local, redis_params))
+  end
+
+  local function add_to_global_replies_set(recipients, sender_key, global_key, task_time_str)
+    lua_util.debugm(N, task, 'Adding recipients %s to global replies set', recipients)
+
+    local function zadd_global_set_cb(err, data)
+      if err ~= nil then
+        rspamd_logger.errx(task, 'failed to add recipients %s to global replies set with error: %s', recipients, err)
+        return
+      end
+      rspamd_logger.infox(task, 'added recipients %s to global set', recipients)
+    end
+    for _, rcpt in ipairs(recipients) do
+      lua_redis.exec_redis_script(script_ids[2],
+              { task=task, is_write = true },
+              zadd_global_set_cb,
+              { global_key },
+              { task_time_str, rcpt })
+    end
+  end
+
+  local function update_global_replies_set(recipients, sender_key, global_key, task_time)
+    local function redis_zrange_cb(err, data)
+      if err ~= nil then
+        rspamd_logger.errx(task,
+                'redis_zrange_cb error when reading zrange withscores from global replies set with error: %s', err)
+        return
+      end
+      local last_score = tonumber(data[#data])
+      lua_util.debugm(N, task, 'last score %s of global replies set was received', last_score)
+
+      -- if last score wasn't found
+      if last_score == nil then
+        lua_util.debugm(N, task,
+                'have not found any senders in global replies set, considering last score as 0')
+        last_score = 0
+      end
+
+      -- updating params considering last score of existing sender
+      task_time = task_time + last_score
+
+      add_to_global_replies_set(recipients, sender_key, global_key, tostring(task_time))
+    end
+
+    lua_util.debugm(N, task, 'Getting recipients withscores from global replies set to get last score')
+
+    -- getting scores of recipients in global replies set
+    lua_redis.redis_make_request(task,
+            redis_params,
+            sender_key,
+            false,
+            redis_zrange_cb,
+            'ZRANGE',
+            {global_key, '0', '-1', 'WITHSCORES'}
```

Comment:
I don't understand this logic as well: why do we get **all** the keys for each check?